<a href="https://colab.research.google.com/github/kingbughk/Datathon-2021/blob/main/AF.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Setup the connections


In [ ]:
#sets the project id
PROJECT_ID = "" #@param {type:"string"}

In [ ]:
import os
from google.colab import auth

#sets dateset
DATASET_PROJECT_ID = 'amsterdamumcdb-data'
DATASET_ID = 'ams102'
LOCATION = 'eu'

#all libraries check this environment variable, so set it:
os.environ["GOOGLE_CLOUD_PROJECT"] = PROJECT_ID

auth.authenticate_user()
print('Authenticated')

Authenticated


##Import the query tools

In [ ]:
from google.cloud.bigquery import magics
from google.cloud import bigquery

#sets the default query job configuration
def_config = bigquery.job.QueryJobConfig(default_dataset=DATASET_PROJECT_ID + "." + DATASET_ID)
magics.context.default_query_job_config = def_config

#sets client options job configuration
client_options = {}
client_options['location'] = LOCATION
magics.context.bigquery_client_options = client_options

#get the amsterdamumcdb package from PiPy repository for use in Colab
!pip install amsterdamumcdb
import amsterdamumcdb as adb

##Get data involving coded with AF

In [ ]:
%%bigquery af_admissions
    SELECT
      admissionid
      ,MIN(measuredat) AS first_af_time
    FROM listitems
    WHERE 
      listitems.itemid = 6671	--Hartritme
      AND listitems.valueid = 13	--Atr fib
    GROUP BY
      admissionid


#Number of admissions with AF anytime

In [ ]:
af_admissions.count()

admissionid      4086
first_af_time    4086
dtype: int64

In [ ]:
af_admissions.head(10)


admissionid  first_af_time
0           93   208733520000
1          160       -1500000
2          254      311280000
3          272       23940000
4          295      141120000
5          299       92940000
6          337    53974200000
7          375      425940000
8          423     1937280000
9          481       20400000

#Examine the record of rhythm of admissionid=295 and test linking 2 data tables

In [ ]:
%%bigquery rhythm_of_295
    SELECT
      listitems.admissionid
      ,listitems.measuredat/1000/60/60 AS time_of_record -- converted to hours
      ,listitems.valueid 
      ,admissions.gender
    FROM listitems, admissions
    WHERE 
      itemid = 6671	--Hartritme
      AND listitems.admissionid = 295
      AND listitems.admissionid=admissions.admissionid
    ORDER BY time_of_record

In [ ]:
rhythm_of_295

admissionid  time_of_record  valueid gender
0           295        3.033333       20  Vrouw
1           295        4.033333       20  Vrouw
2           295        5.033333       20  Vrouw
3           295        6.033333       20  Vrouw
4           295        7.033333        1  Vrouw
..          ...             ...      ...    ...
81          295       91.033333       13  Vrouw
82          295       92.033333       13  Vrouw
83          295       93.033333       13  Vrouw
84          295       94.033333       13  Vrouw
85          295       95.033333       13  Vrouw

[86 rows x 4 columns]

##Examine the types of possible rhythms

In [ ]:
%%bigquery type_of_rhythms
    SELECT DISTINCT
      listitems.valueid
      , listitems.value
    FROM listitems
    WHERE 
      itemid = 6671	--Hartritme

In [ ]:
type_of_rhythms

valueid         value
0        33     P.M-ritme
1        20     Sinus bra
2        14      Atr flut
3         2     AVblock 1
4        17     Sinus ari
5        22           SVT
6         4     AVblock 3
7        25            VT
8        15         Junct
9        18     Sinus arr
10        3     AVblock 2
11       10       ST-depr
12       34     AV-Nodaal
13       12     Atr Tachy
14       26          >QRS
15        9     PAT-block
16       32   >ST-segment
17       35  Ventriculair
18        6   AVblock 3:1
19        8    Compl. blk
20       31   >QT-segment
21        5   AVblock 2:1
22       23     Ventr arr
23       11     Asystolie
24       24            VF
25       16           PAT
26       19     Sinus blk
27        7   AVBlock 4:1
28       30   <ST-segment
29        1           NSR
30       13       Atr fib
31       21     Sinus Tac

## Examine the possible rhythms from dictionary

In [ ]:
dictionary = adb.get_dictionary()
dictionary[(dictionary["itemid"]==6671)]

itemid       item       item_en  ...      table    count count_validated
2359    6671  Hartritme  heart rhythm  ...  listitems  1203779       1203779.0
2360    6671  Hartritme  heart rhythm  ...  listitems    20170         20170.0
2361    6671  Hartritme  heart rhythm  ...  listitems     1305          1305.0
2362    6671  Hartritme  heart rhythm  ...  listitems     1198          1198.0
2363    6671  Hartritme  heart rhythm  ...  listitems      131           131.0
2364    6671  Hartritme  heart rhythm  ...  listitems       49            49.0
2365    6671  Hartritme  heart rhythm  ...  listitems       19            19.0
2366    6671  Hartritme  heart rhythm  ...  listitems      182           182.0
2367    6671  Hartritme  heart rhythm  ...  listitems       76            76.0
2368    6671  Hartritme  heart rhythm  ...  listitems      115           115.0
2369    6671  Hartritme  heart rhythm  ...  listitems      192           192.0
2370    6671  Hartritme  heart rhythm  ...  listitems     1556          1556.0
2371    6671  Hartritme  heart rhythm  ...  listitems   207629        207629.0
2372    6671  Hartritme  heart rhythm  ...  listitems    14186         14186.0
2373    6671  Hartritme  heart rhythm  ...  listitems     2568          2568.0
2374    6671  Hartritme  heart rhythm  ...  listitems      140           140.0
2375    6671  Hartritme  heart rhythm  ...  listitems     4046          4046.0
2376    6671  Hartritme  heart rhythm  ...  listitems      553           553.0
2377    6671  Hartritme  heart rhythm  ...  listitems       68            68.0
2378    6671  Hartritme  heart rhythm  ...  listitems    93519         93519.0
2379    6671  Hartritme  heart rhythm  ...  listitems   394670        394670.0
2380    6671  Hartritme  heart rhythm  ...  listitems     1967          1967.0
2381    6671  Hartritme  heart rhythm  ...  listitems      127           127.0
2382    6671  Hartritme  heart rhythm  ...  listitems      161           161.0
2383    6671  Hartritme  heart rhythm  ...  listitems     2357          2357.0
2384    6671  Hartritme  heart rhythm  ...  listitems     1080          1080.0
2385    6671  Hartritme  heart rhythm  ...  listitems       12            12.0
2386    6671  Hartritme  heart rhythm  ...  listitems      184           184.0
2387    6671  Hartritme  heart rhythm  ...  listitems       71            71.0
2388    6671  Hartritme  heart rhythm  ...  listitems    57210         57210.0
2389    6671  Hartritme  heart rhythm  ...  listitems     3075          3075.0
2390    6671  Hartritme  heart rhythm  ...  listitems      811           811.0

[32 rows x 19 columns]

#Get Rhythm on Admission
##Criteria:

1. First entry made within 1 hour of admission
2. 2 or more entries made within 3 hours of admission
3. All rhythm registered are the same (NSR, Sinus Brady and Sinus Tachy are considered the same)

In [ ]:
%%bigquery rhythm_on_admission
SELECT admissionid, min_value AS admit_rhythm, num_of_entries, admittedat, first_entry_time FROM
(
    SELECT DISTINCT
      listitems.admissionid
      , MIN(CASE WHEN listitems.valueid=1  -- NSR
              OR listitems.valueid=20 -- Sinus Brady
              OR listitems.valueid=21 -- Sinus Tachy
              THEN 1 ELSE listitems.valueid END) AS min_value -- map all sinus rhythm to NSR
      , MAX(CASE WHEN listitems.valueid=1  -- NSR
              OR listitems.valueid=20 -- Sinus Brady
              OR listitems.valueid=21 -- Sinus Tachy
              THEN 1 ELSE listitems.valueid END) AS max_value -- map all sinus rhythm to NSR
      , count(listitems.admissionid) AS num_of_entries
      , admissions.admittedat
      , MIN(listitems.measuredat) AS first_entry_time
    FROM listitems, admissions
    WHERE 
      listitems.itemid = 6671	--Hartritme
      AND listitems.admissionid=admissions.admissionid
      AND admissions.admittedat + 3*60*60*1000 >= listitems.measuredat -- get all entries within 3 hours of admission
      AND admissions.admittedat <= listitems.measuredat -- get all entries within 3 hours of admission
    GROUP BY listitems.admissionid, admissions.admittedat
)
WHERE min_value=max_value -- same rhythm within first 3 hours
      AND num_of_entries > 1 -- more tan 1 entry of rhythm
      AND admittedat + 1*60*60*1000 >= first_entry_time -- first entry within first hour of admission


In [ ]:
rhythm_on_admission.count()

admissionid         12405
admit_rhythm        12405
num_of_entries      12405
admittedat          12405
first_entry_time    12405
dtype: int64

In [ ]:
rhythm_on_admission['admit_rhythm'].value_counts()

1     11283
13      758
33      270
14       27
2        23
17       13
34        9
15        6
22        4
35        3
3         2
12        2
26        1
4         1
25        1
5         1
23        1
Name: admit_rhythm, dtype: int64

In [ ]:
rhythm_on_admission['num_of_entries'].value_counts()

3     8110
2     3001
4     1110
5      140
6       24
7        8
8        7
9        4
13       1
Name: num_of_entries, dtype: int64

##Out of those admitted with SR, find who developed AF

In [ ]:
%%bigquery new_af
SELECT listitems.admissionid, MIN(listitems.measuredat) AS first_AF_time, (MIN(listitems.measuredat) - admissions.admittedat)/1000/60/60 AS time_to_AF

FROM listitems, admissions
WHERE listitems.valueid=13 -- AF
      AND admissions.admissionid=listitems.admissionid
      AND listitems.measuredat > admissions.admittedat
      AND listitems.admissionid IN
                  (SELECT admissionid FROM -- a table with sinus on admission
                      (
                        SELECT admissionid, min_value AS admit_rhythm, num_of_entries, admittedat, first_entry_time FROM
                        (
                            SELECT DISTINCT
                              listitems.admissionid
                              , MIN(CASE WHEN listitems.valueid=1  -- NSR
                                      OR listitems.valueid=20 -- Sinus Brady
                                      OR listitems.valueid=21 -- Sinus Tachy
                                      THEN 1 ELSE listitems.valueid END) AS min_value -- map all sinus rhythm to NSR
                              , MAX(CASE WHEN listitems.valueid=1  -- NSR
                                      OR listitems.valueid=20 -- Sinus Brady
                                      OR listitems.valueid=21 -- Sinus Tachy
                                      THEN 1 ELSE listitems.valueid END) AS max_value -- map all sinus rhythm to NSR
                              , count(listitems.admissionid) AS num_of_entries
                              , admissions.admittedat
                              , MIN(listitems.measuredat) AS first_entry_time
                            FROM listitems, admissions
                            WHERE 
                              listitems.itemid = 6671	--Hartritme
                              AND listitems.admissionid=admissions.admissionid
                              AND admissions.admittedat + 3*60*60*1000 >= listitems.measuredat -- get all entries within 3 hours of admission
                              AND admissions.admittedat <= listitems.measuredat -- get all entries within 3 hours of admission
                            GROUP BY listitems.admissionid, admissions.admittedat
                        )
                        WHERE min_value=1  -- same SR within first 3 hours
                              AND max_value=1 -- same SR within first 3 hours
                              AND num_of_entries > 1 -- more tan 1 entry of rhythm
                              AND admittedat + 1*60*60*1000 >= first_entry_time -- first entry within first hour of admission
                      )
                  )
GROUP BY listitems.admissionid, admissions.admittedat

#Number of new AF (Sinus rhythm on admission)
I'm a bit surprise that out of the 11283 patients admitted with sinus rhythm, 66% (7495/11283) have at least 1 label of AF in their ICU stay.

In [ ]:
new_af.count()

admissionid      7495
first_AF_time    7495
time_to_AF       7495
dtype: int64

In [ ]:
new_af['time_to_AF'].quantile([.05,.25,.5,.75,.95])

0.05     0.216667
0.25     1.050000
0.50     3.583333
0.75    15.775000
0.95    73.926667
Name: time_to_AF, dtype: float64